universal.c

```diff
@@ -593,10 +593,13 @@ XS(XS_utf8_upgrade)
         croak_xs_usage(cv, "sv");
     else {
         SV * const sv = ST(0);
-        STRLEN	RETVAL;
+        STRLEN	RETVAL = 0;
         dXSTARG;
 
-        RETVAL = sv_utf8_upgrade(sv);
+        if (LIKELY(sv && SvANY(sv))) {
```

Comment:
You need to SvGETMAGIC() before checking flags.

You can then use sv_utf8_upgrade_nomg() to avoid double get magic.